src/celeritas/io/EventWriter.cc

```diff
@@ -0,0 +1,189 @@
+#include <HepMC3/WriterAscii.h>
+#include <HepMC3/WriterAsciiHepMC2.h>
+#include <HepMC3/WriterHEPEVT.h>
+
+#include "corecel/io/EnumStringMapper.hh"
+#include "corecel/io/Join.hh"
+#include "corecel/io/Logger.hh"
+#include "corecel/io/ScopedTimeAndRedirect.hh"
+#include "corecel/io/StringEnumMapper.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/phys/Primary.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+EventWriter::Format filename_to_format(std::string const& filename)
+{
+    std::string ext;
+    auto pos = filename.rfind('.');
+    if (pos != std::string::npos)
+    {
+        ext = filename.substr(pos + 1);
+    }
+
+    static auto const from_string
+        = StringEnumMapper<EventWriter::Format>::from_cstring_func(
+            to_cstring, "event file format");
+    return from_string(ext);
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct by parsing the extension.
+ */
+EventWriter::EventWriter(std::string const& filename, SPConstParticles params)
+    : EventWriter{filename, params, filename_to_format(filename)}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a filename, particle data, and output format.
+ */
+EventWriter::EventWriter(std::string const& filename,
+                         SPConstParticles particles,
+                         Format fmt)
+    : particles_(particles)
+{
+    CELER_EXPECT(!filename.empty());
+    CELER_EXPECT(particles_);
+
+    CELER_LOG(info) << "Creating " << to_cstring(fmt) << " event file at "
+                    << filename;
+    ScopedTimeAndRedirect temp_{"HepMC3"};
+
+    if (fmt != Format::hepmc3)
+    {
+        CELER_LOG(warning) << "Writing to a event file format other than "
+                              "hepmc3 may result in an unreadable file";
+    }
+
+    writer_ = [&]() -> std::shared_ptr<HepMC3::Writer> {
+        switch (fmt)
+        {
+            case Format::hepevt:
+                return std::make_shared<HepMC3::WriterHEPEVT>(filename);
+            case Format::hepmc2:
+                return std::make_shared<HepMC3::WriterAsciiHepMC2>(filename);
+            case Format::hepmc3:
+                return std::make_shared<HepMC3::WriterAscii>(filename);
+            default:
+                CELER_ASSERT_UNREACHABLE();
+        }
+    }();
+
+    CELER_ENSURE(writer_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write all primaries from a single event.
+ */
+void EventWriter::operator()(argument_type primaries)
+{
+    CELER_EXPECT(!primaries.empty());
+
+    std::set<EventId::size_type> mismatched_events;
+
+    HepMC3::GenEvent evt;
+    evt.set_units(HepMC3::Units::MEV, HepMC3::Units::CM);
```

Comment:
Since this is a new empty event (i.e. it doesn't already have particles in it that need to be converted), we could also do:
```suggestion
    HepMC3::GenEvent evt(HepMC3::Units::MEV, HepMC3::Units::CM);
```